### PR TITLE
Implements vSHARP model (2D)

### PR DESCRIPTION
**Pull request by georgeyiasemis**
* This PR introduces the variable Splitting Half-quadratic ADMM algorithm for Reconstruction of inverse-Problems (vSHARP) model as presented in https://arxiv.org/abs/2309.09954.
* It includes experiment files as presented in the paper for the fastmri prostate dataset.
* It implements tests for vSHARP
* Contains minor changes that allow None for masking function.
* Updates main README file.

### CODE REVIEW

File: direct/nn/vsharp/vsharp.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
+# Copyright (c) DIRECT Contributors
+
+"""direct.nn.vsharp.vsharp module."""
+
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp.py:4]: It's good to be a little bit more explicit, this ends up in the documentation

File: direct/nn/vsharp/vsharp.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        channels: tuple[int, ...],
+        dilations: tuple[int, ...],
+        multiscale_depth: int = 1,
+        activation: ActivationType = ActivationType.PRELU,
+    ):
+        """Inits :class:`LagrangeMultipliersInitializer`.
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp.py:33]: -> None

File: direct/nn/vsharp/vsharp.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        initializer_multiscale : int
+            Number of multiscale features to include in the  Lagrange multiplier initializer output. Default: 1.
+        initializer_activation : ActivationType
+            Activation type for the Lagrange multiplier initializer. Default: ActivationType.PRELU.
+        auxiliary_steps : int
+            Number of auxiliary steps to output. Can be -1 or a positive integer lower or equal to `num_steps`.
+            If -1, it uses all steps. If I, the last I steps will be used.
+        **kwargs: Additional keyword arguments.
+        """
+        # pylint: disable=too-many-locals
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp.py:189]: Perhaps explain what can be in the kwargs, and why precisely can't you put these s explicit parameters?

File: direct/nn/vsharp/vsharp_engine.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        cfg: BaseConfig,
+        model: nn.Module,
+        device: str,
+        forward_operator: Optional[Callable] = None,
+        backward_operator: Optional[Callable] = None,
+        mixed_precision: bool = False,
+        **models: nn.Module,
+    ):
+        """Inits :class:`VSharpNetEngine`.
+
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp_engine.py:33]: -> None

File: direct/nn/vsharp/vsharp_engine.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        forward_operator: Callable, optional
+            The forward operator. Default: None.
+        backward_operator: Callable, optional
+            The backward operator. Default: None.
+        mixed_precision: bool
+            Use mixed precision. Default: False.
+        **models: nn.Module
+            Additional models.
+        """
+        super().__init__(
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp_engine.py:51]: Can you explain this a bit more elaborately?

File: direct/nn/vsharp/vsharp_engine.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            mixed_precision=mixed_precision,
+            **models,
+        )
+
+    def _do_iteration(
+        self,
+        data: Dict[str, Any],
+        loss_fns: Optional[Dict[str, Callable]] = None,
+        regularizer_fns: Optional[Dict[str, Callable]] = None,
+    ) -> DoIterationOutput:
```
Comment (jonasteuwen) [inline direct/nn/vsharp/vsharp_engine.py:66]: I think modern typing allows to use 'dict' rather than 'Dict'

File: direct/train.py [hunk 7]
```diff
@@ -75,11 +75,13 @@ def get_root_of_file(filename: PathOrString):
 
 
 def build_transforms_from_environment(env, dataset_config: DictConfig) -> Callable:
+    masking = dataset_config.transforms.masking  # Masking func can be None
+    mask_func = None if masking is None else build_masking_function(**masking)
```
Comment (jonasteuwen) [inline direct/train.py:79]: Perhaps some logging would be useful here, I can imagine a small typo would lead to masking = None, a log is helpful in auditing

**Review by jonasteuwen — COMMENTED** [review]
Looks good, only a few minor comments

**Review by jonasteuwen — APPROVED** [review]
LGTM
